File: test/data/database/db_schema_contract_test.dart
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:koru/data/database/app_database.dart';

/// ARCH-04 — Contract test CROSS-RUNTIME (lato Dart).
///
/// Koru ha DUE runtime sullo stesso file `koru.db`: Drift (Dart) possiede lo
/// schema; il lato nativo Kotlin (`android/.../db/NativeDatabase.kt`) lo legge
/// con SQL grezzo e nomi hardcoded. Una rename/drop di colonna lato Drift
/// compila ma rompe l'enforcement nativo SILENZIOSAMENTE a runtime.
///
/// [kNativeSchemaContract] è la mappa (tabella → colonne) di cui il nativo
/// dipende — il MIRROR Dart di `DbSchema.kt`. Questo test apre lo schema VIVO
/// di Drift (`createAll()` via `AppDatabase.forTesting`), interroga
/// `PRAGMA table_info` per ogni tabella del contratto, e asserisce che ogni
/// colonna richiesta dal nativo esista DAVVERO nello schema Drift corrente.
///
/// → Se Drift rinomina/elimina una colonna su cui il nativo si appoggia, questo
///   test FALLISCE. È il pezzo che `DbSchemaContractTest.kt` (Kotlin) non può
///   coprire: quello verifica NativeDatabase vs DbSchema, questo verifica
///   DbSchema (replicato qui) vs lo schema reale di Drift.
///
/// LIMITE ONESTO (no codegen): [kNativeSchemaContract] e `DbSchema.kt` sono
/// mantenuti a mano e DEVONO restare in sync — SONO il contratto cross-runtime.
/// Questo test NON deriva la mappa dallo schema Drift né dal Kotlin: garantisce
/// solo che, DATO il contratto dichiarato, Drift lo soddisfi. Un rename
/// coordinato richiede di aggiornare ENTRAMBE le liste (qui + `DbSchema.kt`).
/// Un guard testuale sotto verifica almeno che le due liste di TABELLE non
/// divergano per dimenticanza grossolana.
///
/// MUST STAY IN SYNC WITH: android/app/src/main/kotlin/com/dev/koru/db/DbSchema.kt
const Map<String, List<String>> kNativeSchemaContract = {
  // profiles — getEnabledProfiles + join in getAllWebsiteRulesForEnabledProfiles
  'profiles': [
    'id',
    'title',
    'type_combinations',
    'on_conditions',
    'operator',
    'day_flags',
    'block_notifications',
    'block_launch',
    'is_enabled',
    'is_locked',
    'on_until',
    'locked_until',
    'paused_until',
    'blocking_mode',
    'block_unsupported_browsers',
    'block_adult_content',
    'color_hex',
    'emoji',
  ],
  // app_profile_relations — getAppRelationsForProfile
  'app_profile_relations': [
    'package_name',
    'profile_id',
    'is_enabled',
    'overlay_config_json',
    'blocked_sections_json',
  ],
  // intervals — getIntervalsForProfile
  'intervals': [
    'id',
    'profile_id',
    'from_minutes',
    'to_minutes',
    'is_enabled',
  ],
  // usage_limits — getUsageLimitsForProfile + updateUsedCount
  'usage_limits': [
    'id',
    'profile_id',
    'period_type',
    'limit_type',
    'last_reset_time',
    'allowed_count',
    'used_count',
  ],
  // website_rules — getWebsiteRulesForProfile + getAllWebsiteRulesForEnabledProfiles
  'website_rules': [
    'id',
    'profile_id',
    'name',
    'blocking_type',
    'is_anywhere_in_url',
    'is_enabled',
  ],
  // wifi_networks — getWifiSsidsByProfile
  'wifi_networks': [
    'profile_id',
    'ssid',
  ],
  // adult_content_sites — isAdultContentSite
  'adult_content_sites': [
    'domain',
  ],
  // block_sessions — insertBlockSession
  'block_sessions': [
    'name',
    'timestamp',
  ],
  // restricted_access_events — insertRestrictedAccessEvent
  'restricted_access_events': [
    'occurred_at',
    'day_start_date',
    'package_name',
    'event_type',
    'restriction_type',
  ],
  // intention_usage_events — insertIntentionEvent
  'intention_usage_events': [
    'occurred_at',
    'day_start_date',
    'package_name',
    'intention_name',
  ],
  // focus_usage_events — insertFocusUsageEvent
  'focus_usage_events': [
    'occurred_at',
    'day_start_date',
    'duration_in_ms',
  ],
};

void main() {
  group('DB schema cross-runtime contract (ARCH-04)', () {
    late AppDatabase db;

    setUp(() {
      // forTesting + NativeDatabase.memory() materializza lo schema reale via
      // MigrationStrategy.onCreate (createAll) — è esattamente lo schema che
      // Drift produrrebbe on-device alla versione corrente.
      db = AppDatabase.forTesting(NativeDatabase.memory());
    });

    tearDown(() async {
      await db.close();
    });

    /// Legge i nomi colonna reali di [table] dallo schema Drift vivo.
    Future<Set<String>> liveColumns(String table) async {
      // PRAGMA table_info(<table>) → una riga per colonna, campo `name`.
      final rows = await db.customSelect('PRAGMA table_info($table)').get();
      return rows.map((r) => r.read<String>('name')).toSet();
    }

    test('ogni tabella del contratto esiste nello schema Drift', () async {
      final names = await db
          .customSelect("SELECT name FROM sqlite_master WHERE type = 'table'")
          .get();
      final liveTables = names.map((r) => r.read<String>('name')).toSet();
      for (final table in kNativeSchemaContract.keys) {
        expect(
          liveTables,
          contains(table),
          reason: 'Il nativo (DbSchema.kt) dipende dalla tabella "$table" ma '
              'non esiste piu\' nello schema Drift. Se l\'hai rinominata/rimossa '
              'aggiorna NativeDatabase.kt + DbSchema.kt + questa mappa.',
        );
      }
    });

    test('ogni colonna richiesta dal nativo esiste nella sua tabella Drift',
        () async {
      final missing = <String>[];
      for (final entry in kNativeSchemaContract.entries) {
        final table = entry.key;
        final live = await liveColumns(table);
        for (final col in entry.value) {
          if (!live.contains(col)) {
            missing.add('$table.$col');
          }
        }
      }
      expect(
        missing,
        isEmpty,
        reason: 'Colonne richieste dal nativo (DbSchema.kt) ma assenti dallo '
            'schema Drift corrente: $missing. Una rawQuery nativa su queste '
            'fallirebbe a runtime ⇒ enforcement spento silenziosamente. '
            'Riallinea NativeDatabase.kt + DbSchema.kt + kNativeSchemaContract '
            '(e aggiungi la migrazione Drift se e\' un rename).',
      );
    });

    test(
        'sanity: una colonna inventata NON e\' nello schema (il PRAGMA discrimina)',
        () async {
      // Guard meta: se il PRAGMA tornasse sempre vuoto/tutto, il test sopra
      // sarebbe un falso-verde. Verifichiamo che discrimini davvero.
      final live = await liveColumns('profiles');
      expect(live, contains('paused_until'));
      expect(live, isNot(contains('paused_until_does_not_exist')));
    });
  });
}

File: android/app/src/test/kotlin/com/dev/koru/db/DbSchemaContractTest.kt
package com.dev.koru.db

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import java.io.File
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * ARCH-04 — Contract test INTERNO al nativo.
 *
 * Costruisce un DB SQLite con ESATTAMENTE le tabelle/colonne dichiarate in
 * [DbSchema] (un proxy minimale dello schema Drift, costruito SOLO da [DbSchema]
 * così non duplica nomi a mano) ed esegue OGNI read query di [NativeDatabase]
 * contro di esso. Se una query referenzia una colonna/tabella non presente in
 * [DbSchema] (un typo, o una query nuova che dimentica di aggiungere la sua
 * colonna a [DbSchema]) il driver SQLite lancia "no such column/table" e il
 * test fallisce.
 *
 * → Cattura il "drift INTERNO" (NativeDatabase che diverge da DbSchema) a
 *   test-time invece che silenziosamente a runtime (rawQuery fallita ⇒
 *   risultato vuoto ⇒ enforcement spento). Il "drift CROSS-RUNTIME" (Drift che
 *   diverge da DbSchema) è coperto dal test Dart gemello.
 *
 * Robolectric fornisce un'implementazione SQLite reale in-JVM, quindi
 * [NativeDatabase] gira esattamente come on-device, incluso il path
 * [NativeDatabase.findDbFile] (il file finisce in `context.filesDir`).
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [33])
class DbSchemaContractTest {

    private val ctx: Context get() = ApplicationProvider.getApplicationContext()
    private val profileId = 1

    @Before
    fun setUp() {
        // Chiudi eventuale handle residuo da un test precedente, poi crea da
        // zero il file `koru.db` dove findDbFile lo cerca (filesDir).
        NativeDatabase.close()
        val dbFile = File(ctx.filesDir, DbSchema.DB_NAME)
        if (dbFile.exists()) dbFile.delete()
        createSchemaFromDbSchema(dbFile)
    }

    @After
    fun tearDown() {
        NativeDatabase.close()
        File(ctx.filesDir, DbSchema.DB_NAME).delete()
    }

    // ─── read queries: nessuna deve lanciare "no such column/table" ──────────

    @Test
    fun getEnabledProfiles_matchesSchema() {
        // Non asseriamo sul contenuto (il DB è vuoto): l'asserzione è "non
        // lancia". Una colonna mancante in DbSchema farebbe esplodere qui.
        assertThat(NativeDatabase.getEnabledProfiles(ctx)).isEmpty()
    }

    @Test
    fun getAppRelationsForProfile_matchesSchema() {
        assertThat(NativeDatabase.getAppRelationsForProfile(ctx, profileId)).isEmpty()
    }

    @Test
    fun getIntervalsForProfile_matchesSchema() {
        assertThat(NativeDatabase.getIntervalsForProfile(ctx, profileId)).isEmpty()
    }

    @Test
    fun getUsageLimitsForProfile_matchesSchema() {
        assertThat(NativeDatabase.getUsageLimitsForProfile(ctx, profileId)).isEmpty()
    }

    @Test
    fun getWifiSsidsByProfile_matchesSchema() {
        assertThat(NativeDatabase.getWifiSsidsByProfile(ctx)).isEmpty()
    }

    @Test
    fun getWebsiteRulesForProfile_matchesSchema() {
        assertThat(NativeDatabase.getWebsiteRulesForProfile(ctx, profileId)).isEmpty()
    }

    @Test
    fun getAllWebsiteRulesForEnabledProfiles_matchesSchema() {
        // Esercita anche il JOIN website_rules × profiles (alias w/p): se una
        // delle due tabelle perde una colonna usata nel join/WHERE, fallisce.
        assertThat(NativeDatabase.getAllWebsiteRulesForEnabledProfiles(ctx)).isEmpty()
    }

    @Test
    fun isAdultContentSite_matchesSchema() {
        assertThat(NativeDatabase.isAdultContentSite(ctx, "example.com")).isFalse()
    }

    // ─── write queries (log/usage): stessi nomi colonna, stesso contratto ────

    @Test
    fun writeQueries_matchSchema() {
        val now = System.currentTimeMillis()
        // updateUsedCount tocca usage_limits (UPDATE ... WHERE id = ?).
        NativeDatabase.updateUsedCount(ctx, limitId = 1, usedCount = 5L)
        NativeDatabase.insertBlockSession(ctx, "com.foo", now)
        NativeDatabase.insertRestrictedAccessEvent(
            ctx, "com.foo", eventType = 0, restrictionType = 0, timestamp = now,
        )
        NativeDatabase.insertIntentionEvent(ctx, "com.foo", "Work", now)
        NativeDatabase.insertFocusUsageEvent(ctx, durationMs = 1_000L, timestamp = now)

        // Verifica che gli INSERT siano davvero atterrati nelle colonne giuste
        // (non solo "non ha lanciato"): un INSERT con colonna sbagliata
        // fallirebbe sopra, ma confermiamo anche il round-trip.
        countRows(DbSchema.BlockSessions.TABLE).let { assertThat(it).isEqualTo(1) }
        countRows(DbSchema.RestrictedAccessEvents.TABLE).let { assertThat(it).isEqualTo(1) }
        countRows(DbSchema.IntentionUsageEvents.TABLE).let { assertThat(it).isEqualTo(1) }
        countRows(DbSchema.FocusUsageEvents.TABLE).let { assertThat(it).isEqualTo(1) }
    }

    // ─── helpers ─────────────────────────────────────────────────────────────

    private fun countRows(table: String): Int {
        val db = NativeDatabase.open(ctx)
        db.rawQuery("SELECT COUNT(*) FROM $table", null).use { c ->
            return if (c.moveToFirst()) c.getInt(0) else -1
        }
    }

    /**
     * Crea le tabelle ESCLUSIVAMENTE dalle costanti di [DbSchema]. Costruendo lo
     * schema da DbSchema (non da literal scritti a mano) garantiamo che il test
     * verifichi NativeDatabase vs DbSchema, non DbSchema vs sé stesso: se
     * NativeDatabase usasse un nome non in DbSchema, la colonna mancherebbe qui
     * e la query esploderebbe. Tipi generici (INTEGER/TEXT) — il contratto è sui
     * NOMI, non sull'affinità.
     */
    private fun createSchemaFromDbSchema(dbFile: File) {
        val db = SQLiteDatabase.openOrCreateDatabase(dbFile, null)
        try {
            db.execSQL(createTable(DbSchema.Profiles.TABLE, profileColumns()))
            db.execSQL(createTable(DbSchema.AppProfileRelations.TABLE, appRelationColumns()))
            db.execSQL(createTable(DbSchema.Intervals.TABLE, intervalColumns()))
            db.execSQL(createTable(DbSchema.UsageLimits.TABLE, usageLimitColumns()))
            db.execSQL(createTable(DbSchema.WebsiteRules.TABLE, websiteRuleColumns()))
            db.execSQL(createTable(DbSchema.WifiNetworks.TABLE, wifiColumns()))
            db.execSQL(createTable(DbSchema.AdultContentSites.TABLE, adultColumns()))
            db.execSQL(createTable(DbSchema.BlockSessions.TABLE, blockSessionColumns()))
            db.execSQL(createTable(DbSchema.RestrictedAccessEvents.TABLE, raeColumns()))
            db.execSQL(createTable(DbSchema.IntentionUsageEvents.TABLE, iueColumns()))
            db.execSQL(createTable(DbSchema.FocusUsageEvents.TABLE, fueColumns()))
        } finally {
            db.close()
        }
    }

    private fun createTable(table: String, columns: List<String>): String =
        "CREATE TABLE $table (" + columns.joinToString(", ") { "$it INTEGER" } + ")"

    private fun profileColumns() = with(DbSchema.Profiles) {
        listOf(
            ID, TITLE, TYPE_COMBINATIONS, ON_CONDITIONS, OPERATOR, DAY_FLAGS,
            BLOCK_NOTIFICATIONS, BLOCK_LAUNCH, IS_ENABLED, IS_LOCKED, ON_UNTIL,
            LOCKED_UNTIL, PAUSED_UNTIL, BLOCKING_MODE, BLOCK_UNSUPPORTED_BROWSERS,
            BLOCK_ADULT_CONTENT, COLOR_HEX, EMOJI,
        )
    }

    private fun appRelationColumns() = with(DbSchema.AppProfileRelations) {
        listOf(PACKAGE_NAME, PROFILE_ID, IS_ENABLED, OVERLAY_CONFIG_JSON, BLOCKED_SECTIONS_JSON)
    }

    private fun intervalColumns() = with(DbSchema.Intervals) {
        listOf(ID, PROFILE_ID, FROM_MINUTES, TO_MINUTES, IS_ENABLED)
    }

    private fun usageLimitColumns() = with(DbSchema.UsageLimits) {
        listOf(ID, PROFILE_ID, PERIOD_TYPE, LIMIT_TYPE, LAST_RESET_TIME, ALLOWED_COUNT, USED_COUNT)
    }

    private fun websiteRuleColumns() = with(DbSchema.WebsiteRules) {
        listOf(ID, PROFILE_ID, NAME, BLOCKING_TYPE, IS_ANYWHERE_IN_URL, IS_ENABLED)
    }

    private fun wifiColumns() = with(DbSchema.WifiNetworks) { listOf(PROFILE_ID, SSID) }

    private fun adultColumns() = with(DbSchema.AdultContentSites) { listOf(DOMAIN) }

    private fun blockSessionColumns() = with(DbSchema.BlockSessions) { listOf(NAME, TIMESTAMP) }

    private fun raeColumns() = with(DbSchema.RestrictedAccessEvents) {
        listOf(OCCURRED_AT, DAY_START_DATE, PACKAGE_NAME, EVENT_TYPE, RESTRICTION_TYPE)
    }

    private fun iueColumns() = with(DbSchema.IntentionUsageEvents) {
        listOf(OCCURRED_AT, DAY_START_DATE, PACKAGE_NAME, INTENTION_NAME)
    }

    private fun fueColumns() = with(DbSchema.FocusUsageEvents) {
        listOf(OCCURRED_AT, DAY_START_DATE, DURATION_IN_MS)
    }
}
